Add spec header with link to open file in editor

panzarino opened the pull request:
<!-- 
Thanks for contributing!
Read our contribution guidelines here: 
https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md 
-->

<!-- Example: "Closes #1234" -->

- Closes #7506 

### User facing changelog

Adds a header above the list of running specs that contains the file name. When the user clicks on the file name, that file will be opened in their editor of choice.

### How has the user experience changed?

<!--
Provide before and after examples of the change.
Screenshots or GIFs are preferred.
-->
Before:
![Screen Shot 2020-05-29 at 1 16 38 PM](https://user-images.githubusercontent.com/7033952/83286959-d0f01d80-a1ae-11ea-9361-4896437554f8.png)

After:
![Screen Shot 2020-05-29 at 1 27 27 PM](https://user-images.githubusercontent.com/7033952/83287757-32fd5280-a1b0-11ea-8967-c6f945f89224.png)

**When clicking Run All Specs**

<img width="827" alt="Screen Shot 2020-06-05 at 1 00 54 PM" src="https://user-images.githubusercontent.com/1271364/83844267-a8c36a00-a72c-11ea-98f6-196dc77025d2.png">


![ezgif-6-3600bc2c0d99](https://user-images.githubusercontent.com/7033952/83287519-d39f4280-a1af-11ea-9a42-3dc71c3bfdc6.gif)

### PR Tasks

<!-- 
These tasks must be completed before a PR is merged.
Delete tasks if they are not applicable. 
-->

- [x] Have tests been added/updated?
- [x] Has the original issue been tagged with a release in ZenHub? <!-- (internal team only)-->
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? https://github.com/cypress-io/cypress-documentation/pull/2852

chrisbreiding commented:
Let me know if you need help debugging those failing e2e tests. Those ones are a little odd and I'm very familiar with them since I wrote them recently for the error improvements work.

panzarino commented:
@chrisbreiding I've fixed all of the errors that you've mentioned above.

- With the last change, it adds the `cypress` prefix to all of the displayed file names, which is now consistent with what is displayed in the error message file names.
- Regarding the change with the special characters, I fixed the display but also noticed that clicking links on file names with special characters does not work either. In addition, when an error occurs in a file like this, there is no clickable stack trace and no code frame at all. I feel like fixing those issues is outside of the scope of this PR.

I took a look at the failing e2e tests for a while but wasn't able to figure out why my changes caused them to time out so any help would be appreciated.

<img width="528" alt="Screen Shot 2020-05-29 at 3 30 31 PM" src="https://user-images.githubusercontent.com/7033952/83300997-f38e3080-a1c6-11ea-861f-15910aee316e.png">
<img width="535" alt="Screen Shot 2020-05-29 at 3 50 16 PM" src="https://user-images.githubusercontent.com/7033952/83300993-f1c46d00-a1c6-11ea-832f-3615c6ff0f96.png">
<img width="519" alt="Screen Shot 2020-05-29 at 4 11 43 PM" src="https://user-images.githubusercontent.com/7033952/83301164-22a4a200-a1c7-11ea-8a51-5c64a6dfdc75.png">

chrisbreiding commented on packages/runner/src/app/app.jsx:
```diff
@@ -22,6 +22,7 @@ class App extends Component {
 
   render () {
     const specPath = this.props.util.absoluteSpecPath(this.props.config)
+    const relativeSpecPath = decodeURI(this.props.util.relativeSpecPath(this.props.config))
```
I'm still seeing the issue with special characters. Can you add an integration test covering this? Have it include `?`, `&`, and `%`, since those are the ones [we have special handling for](https://github.com/cypress-io/cypress/blob/develop/packages/server/lib/util/escape_filename.ts). I think the solution is to use `decodeURIComponent` instead of `decodeURI`.

panzarino replied:
@chrisbreiding  I've made this change but I am unsure of how to proceed regarding tests. Within the existing integration tests, the reporter is rendered directly without being launched from the runner. It seems like this would require an `e2e` test, but I'm pretty unfamiliar with how those work.

chrisbreiding commented:
> I [...] noticed that clicking links on file names with special characters does not work either. In addition, when an error occurs in a file like this, there is no clickable stack trace and no code frame at all. I feel like fixing those issues is outside of the scope of this PR.

Good find. I think this is due to having a space in the name rather than the special characters. At least, it can be reproduced with only a space. You're right that's it's outside the scope of this PR, so I've opened #7553 for tracking that.

panzarino commented:
@chrisbreiding The latest changes should resolve the issues with wrapping. I changed the position to `sticky`, which allows it to stay sticky on scroll and act as if its position is relative before scroll.

I think everything should be good to go now except for the tests regarding special characters, if you would be willing to help out/provide direction for those.

chrisbreiding commented:
@panzarino I [added a test](https://github.com/cypress-io/cypress/pull/7515/commits/eca704d4cfa43d47e990bf623b2d67903d07a504) to cover the special characters. Was able to do it in the reporter tests, but had to get a little tricky. Instead of rendering the reporter separately in the AUT, I put the special characters in the file name and test the actual reporter itself. Can't test `?` though since it's only valid in Linux and would fail elsewhere.

Also, I discovered that GitHub can't handle [file names with those characters](https://github.com/cypress-io/cypress/tree/issue-7506/packages/reporter/cypress/integration/meta_&%_spec.ts) 😂

brian-mann commented on packages/runner/src/lib/util.js:
```diff
@@ -19,6 +15,10 @@ export default {
   },
 
+  relativeSpecPath (config) {
+    return path.relative(config.fileServerFolder, this.absoluteSpecPath(config))
```
The problem with this method is that it is using `this.absoluteSpecPath()` which the implementation written for it **is not** the logic you want. That's generating a valid URL for the purposes of running the spec via URL. The magic `integration` word is used there because in the URL Cypress knows which kind of spec file you're searching for on the file system.

However, for the purposes of what **you want** it is simply just `Cypress.spec.relative` which will always give you the relative path to file on the users hard drive relative from the `projectRoot`. You could chop off the `cypress/integration` part if you'd like using `path.resolve(config.integrationFolder, Cypress.spec.relative)`.

panzarino replied:
@brian-mann Wow, really wish I knew about that earlier... completely simplifies a ton of stuff from this PR. I've now updated the PR to take advantage of `Cypress.spec.relative`.

As far as chopping off `cypress/integration` part, I think it's best to leave it now as it's consistent with what is shown in the error messages and code frame.

brian-mann replied:
agree

bahmutov commented:
I will upgrade cypress-react-unit-test in Cypress monorepo and will show
how to write component tests. This will make testing edge cases like this
much easier



On Tue, Jun 2, 2020 at 9:16 AM Chris Breiding <notifications@github.com>
wrote:

> *@chrisbreiding* approved this pull request.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cypress-io/cypress/pull/7515#pullrequestreview-422663588>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQ4BJX2H7ZIGQ7BVBQYUBTRUT3RXANCNFSM4NNHUROQ>
> .
>


-- 
Dr. Gleb Bahmutov, PhD

Schedule video chat / phone call / meeting with me via
https://calendly.com/bahmutov
gleb.bahmutov@gmail.com @bahmutov <https://twitter.com/@bahmutov>
https://glebbahmutov.com/ https://glebbahmutov.com/blog
https://github.com/bahmutov

chrisbreiding commented:
@bahmutov You're referring to the special characters test? I'm not sure a component test would be as valuable in this case as an e2e test. But for most of the reporter tests, I agree that they could be converted to component tests.

bahmutov commented:
I might have commented on the wrong email - I thought I was replying to error code frames ;)

Sent from my iPhone

> On Jun 2, 2020, at 09:34, Chris Breiding <notifications@github.com> wrote:
> 
> ﻿
> @bahmutov You're referring to the special characters test? I'm not sure a component test would be as valuable in this case as an e2e test. But for most of the reporter tests, I agree that they could be converted to component tests.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

chrisbreiding commented:
@jennifer-shehane Do you think we should add something to the docs about this?

brian-mann commented:
@panzarino branch conflicts, fix, gogogogoog

panzarino commented:
@jennifer-shehane When all specs are running, the header will now display 'All Specs' and not be clickable. I think showing all of the spec names scrolling by is outside of the scope of this PR for now.

I also noticed that @chrisbreiding's test regarding special characters will fail when all tests are run (because the header won't be for that specific file) so maybe that test should be rewritten/removed before these changes are merged. Because it seems like special characters have issues all over, it might be best to create a separate issue and PR to handle everything on that subject.

chrisbreiding commented:
> I also noticed that @chrisbreiding's test regarding special characters will fail when all tests are run (because the header won't be for that specific file) so maybe that test should be rewritten/removed before these changes are merged

I wouldn't worry too much about it. The `All Specs` thing is only in interactive mode is not something we really use (or advocate using). We could update the test to also accept `All Specs`, but I'm not sure it's worth the effort.

> it might be best to create a separate issue and PR to handle everything on that subject

Agree. The issue with special characters when there are no tests is unrelated to this feature, so it should be handled separately.
